File: typr-tester-typr-scala-new/src/scala/adventureworks/public/OpenEnumTest.scala
package adventureworks.public

import adventureworks.public.title.*
import adventureworks.public.title_domain.*
import adventureworks.public.titledperson.*
import adventureworks.{DomainInsertImpl, TestInsert, WithConnection}
import org.junit.Assert.*
import org.junit.Test

import java.util.Random

class OpenEnumTest {
  private val titleRepo = TitleRepoImpl()
  private val titleDomainRepo = TitleDomainRepoImpl()
  private val titledPersonRepo = TitledpersonRepoImpl()

  @Test
  def testOpenEnumPatternMatching(): Unit = {
    WithConnection {
      val testInsert = new TestInsert(new Random(0), new DomainInsertImpl())
      val john = testInsert.publicTitledperson(TitleDomainId.dr, TitleId.dr, "John")

      // Verify pattern matching works with open enums
      val titleValue = john.title
      val titleString = titleValue match {
        case TitleId.dr               => "Known: dr"
        case TitleId.mr               => "Known: mr"
        case TitleId.ms               => "Known: ms"
        case TitleId.phd              => "Known: phd"
        case unknown: TitleId.Unknown => s"Unknown: ${unknown.value}"
      }
      assertEquals("Known: dr", titleString)

      // Test DSL query with joinFk chain
      val found = titledPersonRepo.select
        .joinFk(tp => tp.fkTitle, titleRepo.select.where(t => t.code.in(Array[Object](TitleId.dr), TitleId.pgType)))
        .joinFk(tp_t => tp_t._1.fkTitleDomain, titleDomainRepo.select.where(td => td.code.in(Array[Object](TitleDomainId.dr), TitleDomainId.pgType)))
        .where(tp_t_td => tp_t_td._1._1.name.isEqual("John"))
        .toList
        .headOption

      assertTrue(found.isDefined)
      val result = found.get
      assertEquals(john, result._1._1)
      assertEquals(TitleRow(TitleId.dr), result._1._2)
      assertEquals(TitleDomainRow(TitleDomainId.dr), result._2)
    }
  }
}

File: typr-tester-typr-java/src/java/adventureworks/public_/OpenEnumTest.java
package adventureworks.public_;

import static org.junit.Assert.*;

import adventureworks.DomainInsertImpl;
import adventureworks.TestInsert;
import adventureworks.WithConnection;
import adventureworks.public_.title.*;
import adventureworks.public_.title_domain.*;
import adventureworks.public_.titledperson.*;
import java.util.Random;
import org.junit.Test;

/**
 * Tests for open enum types (sealed interface with Known enum + Unknown record). Equivalent to
 * Scala OpenEnumTest.
 */
public class OpenEnumTest {
  private final TitleRepoImpl titleRepo = new TitleRepoImpl();
  private final TitleDomainRepoImpl titleDomainRepo = new TitleDomainRepoImpl();
  private final TitledpersonRepoImpl titledPersonRepo = new TitledpersonRepoImpl();

  @Test
  public void works() {
    WithConnection.run(
        c -> {
          var testInsert = new TestInsert(new Random(0), new DomainInsertImpl());
          var john =
              testInsert.publicTitledperson(TitleDomainId.Known.dr, TitleId.Known.dr, "John", c);

          // DSL query with joinFk chain
          var found =
              titledPersonRepo
                  .select()
                  .joinFk(
                      tp -> tp.fkTitle(),
                      titleRepo.select().where(t -> t.code().in(TitleId.Known.dr)))
                  .joinFk(
                      tp_t -> tp_t._1().fkTitleDomain(),
                      titleDomainRepo.select().where(td -> td.code().in(TitleDomainId.Known.dr)))
                  .where(tp_t_td -> tp_t_td._1()._1().name().isEqual("John"))
                  .toList(c)
                  .stream()
                  .findFirst();

          assertTrue(found.isPresent());
          var result = found.get();
          // result is ((TitledpersonRow, TitleRow), TitleDomainRow)
          assertEquals(john, result._1()._1());
          assertEquals(new TitleRow(TitleId.Known.dr), result._1()._2());
          assertEquals(new TitleDomainRow(TitleDomainId.Known.dr), result._2());
        });
  }

  private void testDsl(
      TitledpersonRepo titledPersonRepo, TitleRepo titleRepo, TitleDomainRepo titleDomainRepo) {
    WithConnection.run(
        c -> {
          // Insert the titled person with dr/dr/John
          var john =
              titledPersonRepo.insert(
                  new TitledpersonRow(TitleDomainId.Known.dr, TitleId.Known.dr, "John"), c);

          // DSL query with joinFk chain
          var found =
              titledPersonRepo
                  .select()
                  .joinFk(
                      tp -> tp.fkTitle(),
                      titleRepo.select().where(t -> t.code().in(TitleId.Known.dr)))
                  .joinFk(
                      tp_t -> tp_t._1().fkTitleDomain(),
                      titleDomainRepo.select().where(td -> td.code().in(TitleDomainId.Known.dr)))
                  .where(tp_t_td -> tp_t_td._1()._1().name().isEqual("John"))
                  .toList(c)
                  .stream()
                  .findFirst();

          assertTrue(found.isPresent());
          var result = found.get();
          assertEquals(john, result._1()._1());
          assertEquals(new TitleRow(TitleId.Known.dr), result._1()._2());
          assertEquals(new TitleDomainRow(TitleDomainId.Known.dr), result._2());

          // Verify pattern matching works with open enums
          var titleValue = john.title();
          String titleString =
              switch (titleValue) {
                case TitleId.Known known -> "Known: " + known.value();
                case TitleId.Unknown unknown -> "Unknown: " + unknown.value();
              };
          assertEquals("Known: dr", titleString);
        });
  }

  @Test
  public void pg() {
    testDsl(new TitledpersonRepoImpl(), new TitleRepoImpl(), new TitleDomainRepoImpl());
  }

  // Note: inMemory test not implemented because TitledpersonRepoMock is not generated
  // and the existing mocks don't support DSL operations
}
